File: src/main/java/com/wmstein/tourcount/ListSpeciesActivity.java
package com.wmstein.tourcount;

import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ScrollView;

import com.wmstein.tourcount.database.Count;
import com.wmstein.tourcount.database.CountDataSource;
import com.wmstein.tourcount.database.DbHelper;
import com.wmstein.tourcount.database.Head;
import com.wmstein.tourcount.database.HeadDataSource;
import com.wmstein.tourcount.database.Individuals;
import com.wmstein.tourcount.database.IndividualsDataSource;
import com.wmstein.tourcount.database.Section;
import com.wmstein.tourcount.database.SectionDataSource;
import com.wmstein.tourcount.widgets.ListHeadWidget;
import com.wmstein.tourcount.widgets.ListIndividualWidget;
import com.wmstein.tourcount.widgets.ListLineWidget;
import com.wmstein.tourcount.widgets.ListMetaWidget;
import com.wmstein.tourcount.widgets.ListSpRemWidget;
import com.wmstein.tourcount.widgets.ListSpeciesWidget;
import com.wmstein.tourcount.widgets.ListSumWidget;
import com.wmstein.tourcount.widgets.ListTitleWidget;

import java.util.List;

import static java.lang.Math.sqrt;

/****************************************************
 * ListSpeciesActivity shows list of counting results
 * Created by wmstein on 15.03.2016
 */
public class ListSpeciesActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener
{
    private static String TAG = "tourcountListSpeciesAct";
    private TourCountApplication tourCount;
    private SharedPreferences prefs;

    private LinearLayout spec_area;

    // preferences
    private boolean awakePref;
    private String sortPref;
    private boolean screenOrientL; // option for screen orientation

    // the actual data
    private CountDataSource countDataSource;
    private SectionDataSource sectionDataSource;
    private HeadDataSource headDataSource;
    private IndividualsDataSource individualsDataSource;

    private SQLiteDatabase database;
    private DbHelper dbHandler;
    ListSumWidget lsw;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        tourCount = (TourCountApplication) getApplication();
        prefs = TourCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        getPrefs();

        if (screenOrientL)
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        }
        else
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        setContentView(R.layout.activity_list_species);

        countDataSource = new CountDataSource(this);
        sectionDataSource = new SectionDataSource(this);
        headDataSource = new HeadDataSource(this);
        individualsDataSource = new IndividualsDataSource(this);

        ScrollView listSpec_screen = (ScrollView) findViewById(R.id.listSpecScreen);
        assert listSpec_screen != null;
        listSpec_screen.setBackground(tourCount.getBackground());

        //noinspection ConstantConditions
        getSupportActionBar().setTitle(getString(R.string.viewSpecTitle));

        spec_area = (LinearLayout) findViewById(R.id.listSpecLayout);

        if (awakePref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    /*
     * So preferences can be loaded at the start, and also when a change is detected.
     */
    private void getPrefs()
    {
        awakePref = prefs.getBoolean("pref_awake", true);
        sortPref = prefs.getString("pref_sort_sp", "none"); // sorted species list
        screenOrientL = prefs.getBoolean("screen_Orientation", false);
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        if (awakePref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }

        spec_area.removeAllViews();
        loadData();
    }

    // fill ListSpeciesWidget with relevant counts and sections data
    private void loadData()
    {
        int sumsp = 0, sumind = 0, iid = 0;
        double mUncert = 0;
        double longi = 0, lati = 0, heigh = 0, uncer = 0;
        int frst = 0;
        double lo = 0, la = 0, loMin = 0, loMax = 0, laMin = 0, laMax = 0, uc = 0, uncer1 = 0;

        headDataSource.open();
        sectionDataSource.open();

        //load head and meta data
        Head head = headDataSource.getHead();
        Section section = sectionDataSource.getSection();

        // display the list name
        ListTitleWidget elw = new ListTitleWidget(this, null);
        elw.setListTitle(getString(R.string.titleEdit));
        elw.setListName(section.name);
        spec_area.addView(elw);

        // display the list remark
        ListTitleWidget erw = new ListTitleWidget(this, null);
        erw.setListTitle(getString(R.string.notesHere));
        erw.setListName(section.notes);
        spec_area.addView(erw);

        // display the head data
        ListHeadWidget ehw = new ListHeadWidget(this, null);
        ehw.setWidgetLCo(getString(R.string.country));
        ehw.setWidgetLCo1(section.country);
        ehw.setWidgetLName(getString(R.string.inspector));
        ehw.setWidgetLName1(head.observer);
        spec_area.addView(ehw);

        // setup the data sources
        countDataSource.open();
        individualsDataSource.open();
        Individuals individuals;

        // Calculate mean average values for coords and uncertainty
        dbHandler = new DbHelper(this);
        database = dbHandler.getWritableDatabase();
        Cursor curAInd;
        curAInd = database.rawQuery("select * from " + dbHandler.INDIVIDUALS_TABLE, null);
        frst = 0;
        while (curAInd.moveToNext())
        {
            longi = curAInd.getDouble(4);
            lati = curAInd.getDouble(3);
            uncer = Math.rint(curAInd.getDouble(6));

            if (longi != 0) // has coordinates
            {
                //Toast.makeText(getApplicationContext(), longi, Toast.LENGTH_SHORT).show();
                if (MyDebug.LOG)
                    Log.d(TAG, "longi " + longi);
                if (frst == 0)
                {
                    loMin = longi;
                    loMax = longi;
                    laMin = lati;
                    laMax = lati;
                    uncer1 = uncer;
                    frst = 1; // just 1 with coordinates
                }
                else
                {
                    loMin = Math.min(loMin, longi);
                    loMax = Math.max(loMax, longi);
                    laMin = Math.min(laMin, lati);
                    laMax = Math.max(laMax, lati);
                    uncer1 = Math.max(uncer1, uncer);
                }
            }
        }
        curAInd.close();

        lo = (loMax + loMin) / 2;   // average longitude
        la = (laMax + laMin) / 2;   // average latitude

        // Simple distance calculation between 2 coordinates within the temperate zone in meters (Pythagoras):
        //   uc = (((loMax-loMin)*71500)² + ((laMax-laMin)*111300)²)½ 
        uc = sqrt(((Math.pow((loMax - loMin) * 71500, 2)) + (Math.pow((laMax - laMin) * 111300, 2))));
        uc = Math.rint(uc / 2) + 20; // average uncertainty radius + default gps uncertainty
        if (uc <= uncer1)
            uc = uncer1;

        // display the meta data
        ListMetaWidget etw = new ListMetaWidget(this, null);
        etw.setMetaWidget(section);
        etw.setWidget_dla2(la);
        etw.setWidget_dlo2(lo);
        etw.setWidget_muncert2(uc);
        spec_area.addView(etw);

        // load the species data
        List<Count> specs; //List of species
        switch (sortPref)
        {
        case "names_alpha":
            specs = countDataSource.getCntSpeciesSrtName();
            break;
        case "codes":
            specs = countDataSource.getCntSpeciesSrtCode();
            break;
        default:
            specs = countDataSource.getCntSpecies();
            break;
        }

        // calculate the totals
        for (Count spec : specs)
        {
            ListSpeciesWidget widget = new ListSpeciesWidget(this, null);
            widget.setCount(spec);
            int spec_count = widget.getSpec_count(spec);

            sumind = sumind + spec_count; // sum of counted individuals
            sumsp = sumsp + 1;              // sum of counted species
        }

        // display the totals
        lsw = new ListSumWidget(this, null);
        lsw.setSum(sumsp, sumind);
        spec_area.addView(lsw);

        List<Individuals> indivs; // List of individuals

        // display all the counts by adding them to listSpecies layout
        for (Count spec : specs)
        {
            ListSpeciesWidget widget = new ListSpeciesWidget(this, null);
            widget.setCount(spec);
            int spec_count = widget.getSpec_count(spec);
            ListSpRemWidget rwidget = new ListSpRemWidget(this, null);
            rwidget.setCount(spec);
            String tRem = rwidget.getRem(spec);

            // fill widget only for counted species
            if (spec_count > 0)
            {
                spec_area.addView(widget);
                if (!tRem.equals(""))
                {
                    spec_area.addView(rwidget);
                }

                String iName = widget.getSpec_name(spec);
                indivs = individualsDataSource.getIndividualsByName(iName);
                for (Individuals indiv : indivs)
                {
                    ListIndividualWidget iwidget = new ListIndividualWidget(this, null);
                    //load the individuals data
                    iwidget.setIndividual(indiv);
                    spec_area.addView(iwidget);
                }
            }
        }
        ListLineWidget lwidget = new ListLineWidget(this, null);
        spec_area.addView(lwidget);
    }

    @Override
    protected void onPause()
    {
        super.onPause();

        // close the data sources
        headDataSource.close();
        countDataSource.close();
        sectionDataSource.close();

        if (awakePref)
        {
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }


    /***************/
    public void saveAndExit(View view)
    {
        super.finish();
    }

    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        ScrollView listSpec_screen = (ScrollView) findViewById(R.id.listSpecScreen);
        assert listSpec_screen != null;
        listSpec_screen.setBackground(null);
        listSpec_screen.setBackground(tourCount.setBackground());
        getPrefs();
    }

}

File: tourcount/src/main/java/com/wmstein/tourcount/ListSpeciesActivity.kt
package com.wmstein.tourcount

import android.annotation.SuppressLint
import android.database.Cursor
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.ScrollView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import com.wmstein.tourcount.database.Count
import com.wmstein.tourcount.database.CountDataSource
import com.wmstein.tourcount.database.DbHelper
import com.wmstein.tourcount.database.Head
import com.wmstein.tourcount.database.HeadDataSource
import com.wmstein.tourcount.database.Individuals
import com.wmstein.tourcount.database.IndividualsDataSource
import com.wmstein.tourcount.database.SectionDataSource
import com.wmstein.tourcount.widgets.ListIndivNoteWidget
import com.wmstein.tourcount.widgets.ListIndividualWidget
import com.wmstein.tourcount.widgets.ListLineWidget
import com.wmstein.tourcount.widgets.ListLocationWidget
import com.wmstein.tourcount.widgets.ListMetaWidget
import com.wmstein.tourcount.widgets.ListSpeciesWidget
import com.wmstein.tourcount.widgets.ListSumWidget
import com.wmstein.tourcount.widgets.ListTitleWidget
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sqrt

/****************************************************
 * ListSpeciesActivity shows list of counting results
 * Created by wmstein on 2016-03-15,
 * last edited in Java on 2022-05-21,
 * converted to Kotlin on 2023-07-09,
 * last edited on 2024-08-23
 */
class ListSpeciesActivity : AppCompatActivity() {
    private var tourCount: TourCountApplication? = null

    private var specArea: LinearLayout? = null
    var head: Head? = null

    // preferences
    private var prefs = TourCountApplication.getPrefs()
    private var awakePref = false
    private var outPref: String? = null

    // the actual data
    private var countDataSource: CountDataSource? = null
    private var sectionDataSource: SectionDataSource? = null
    private var headDataSource: HeadDataSource? = null
    private var individualsDataSource: IndividualsDataSource? = null
    private var lsw: ListSumWidget? = null

    @SuppressLint("SourceLockedOrientationActivity")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        tourCount = application as TourCountApplication
        awakePref = prefs.getBoolean("pref_awake", true)
        outPref = prefs.getString("pref_sort_output", "names") // sort mode output

        setContentView(R.layout.activity_list_species)
        countDataSource = CountDataSource(this)
        sectionDataSource = SectionDataSource(this)
        headDataSource = HeadDataSource(this)
        individualsDataSource = IndividualsDataSource(this)

        val resultsScreen = findViewById<ScrollView>(R.id.listSpecScreen)
        resultsScreen.background = tourCount!!.background
        supportActionBar!!.title = getString(R.string.viewSpecTitle)
        specArea = findViewById(R.id.listSpecLayout)
        if (awakePref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        // new onBackPressed logic
        if (Build.VERSION.SDK_INT >= 33) {
            onBackPressedDispatcher.addCallback(object :
                OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    headDataSource!!.close()
                    sectionDataSource!!.close()
                    countDataSource!!.close()
                    individualsDataSource!!.close()

                    NavUtils.navigateUpFromSameTask(this@ListSpeciesActivity)
                }
            })
        }
    }

    override fun onResume() {
        super.onResume()
        if (awakePref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }

        // setup the data sources
        headDataSource!!.open()
        sectionDataSource!!.open()
        countDataSource!!.open()
        individualsDataSource!!.open()

        // build Show Results screen
        specArea!!.removeAllViews()
        loadData()
    }

    // fill ListSpeciesWidget with relevant counts and sections data
    private fun loadData() {
        var summf = 0
        var summ = 0
        var sumf = 0
        var sump = 0
        var suml = 0
        var sumo = 0
        var sumsp = 0  // sum of different species
        var sumind = 0 // sum of counted individuals
        var longi: Double
        var lati: Double
        var uncer: Double
        var frst = 0
        val lo: Double
        val la: Double
        var loMin = 0.0
        var loMax = 0.0
        var laMin = 0.0
        var laMax = 0.0
        var uc: Double
        var uncer1 = 0.0

        //load head and meta data from DB
        head = headDataSource!!.head
        val section = sectionDataSource!!.section

        // Build the screen
        // 1. Display list name, observer name and remark by ListTitleWidget
        val ltw = ListTitleWidget(this, null)
        ltw.setListTitle(getString(R.string.titleEdit))
        ltw.setListName(section.name)

        ltw.setWidgetName1(getString(R.string.inspector))
        ltw.setWidgetName2(head!!.observer)

        ltw.setWidgetNotes1(getString(R.string.notesHere))
        ltw.setWidgetNotes2(section.notes)
        specArea!!.addView(ltw)

        // Calculate mean average values for coords and uncertainty
        val dbHandler = DbHelper(this)
        val database = dbHandler.writableDatabase
        val curAInd: Cursor = database.rawQuery("select * from " + DbHelper.INDIVIDUALS_TABLE, null)
        while (curAInd.moveToNext()) {
            longi = curAInd.getDouble(4)
            lati = curAInd.getDouble(3)
            uncer = round(curAInd.getDouble(6))
            if (longi != 0.0) // has coordinates
            {
                if (MyDebug.LOG) Log.d(TAG, "165, longitude: $longi")
                if (frst == 0) {
                    loMin = longi
                    loMax = longi
                    laMin = lati
                    laMax = lati
                    uncer1 = uncer
                    frst = 1 // just 1 with coordinates
                } else {
                    loMin = loMin.coerceAtMost(longi)
                    loMax = loMax.coerceAtLeast(longi)
                    laMin = laMin.coerceAtMost(lati)
                    laMax = laMax.coerceAtLeast(lati)
                    uncer1 = uncer1.coerceAtLeast(uncer)
                }
            }
        }
        curAInd.close()
        dbHandler.close()

        lo = (loMax + loMin) / 2 // average longitude
        la = (laMax + laMin) / 2 // average latitude

        // Simple distance calculation between 2 coordinates within the temperate zone in meters (Pythagoras):
        //   uc = (((loMax-loMin)*71500)² + ((laMax-laMin)*111300)²)½ 
        uc = sqrt(
            ((loMax - loMin) * 71500).pow(2.0) + ((laMax - laMin) * 111300).pow(2.0)
        )
        uc = round(uc / 2) + 20 // average uncertainty radius + default gps uncertainty
        if (uc <= uncer1) uc = uncer1

        // 2. Display the location data
        val llw = ListLocationWidget(this, null)
        llw.setLocationWidget(section)
        llw.setWidgetDla2(la)
        llw.setWidgetDlo2(lo)
        llw.setWidgetMuncert2(uc)
        specArea!!.addView(llw)

        // 3. Display the date, time, temperature, wind and clouds data
        val lmw = ListMetaWidget(this, null)
        lmw.setListMetaWidget(section)
        specArea!!.addView(lmw)

        // load the species data
        val specs: List<Count> // List of sorted species
        if (outPref.equals("names")) {
            // sort criteria are name and code
            specs = countDataSource!!.cntSpeciesSrtName
        } else {
            // sort criteria are name and section
            specs = countDataSource!!.cntSpeciesSrtCode
        }

        // calculate the totals
        var specCntf1i: Int
        var specCntf2i: Int
        var specCntf3i: Int
        var specCntpi: Int
        var specCntli: Int
        var specCntei: Int
        for (spec in specs) {
            val widget = ListSpeciesWidget(this, null)
            widget.setCount(spec)
            specCntf1i = widget.getSpec_countf1i(spec)
            specCntf2i = widget.getSpec_countf2i(spec)
            specCntf3i = widget.getSpec_countf3i(spec)
            specCntpi = widget.getSpec_countpi(spec)
            specCntli = widget.getSpec_countli(spec)
            specCntei = widget.getSpec_countei(spec)
            summf += specCntf1i
            summ += specCntf2i
            sumf += specCntf3i
            sump += specCntpi
            suml += specCntli
            sumo += specCntei
            sumind = (sumind + specCntf1i + specCntf2i + specCntf3i + specCntpi
                    + specCntli + specCntei) // sum of counted individuals
            sumsp += 1 // sum of counted species
        }

        // display line and the totals
        lsw = ListSumWidget(this, null)
        lsw!!.setSum(sumsp, sumind)
        specArea!!.addView(lsw)

        // display all individuals by adding them to listSpecies layout
        var specCnt: Int
        var indivs: List<Individuals> // List of individuals
        var iwidget: ListIndividualWidget

        for (spec in specs) {
            val widget = ListSpeciesWidget(this, null)
            widget.setCount(spec)
            specCntf1i = widget.getSpec_countf1i(spec)
            specCntf2i = widget.getSpec_countf2i(spec)
            specCntf3i = widget.getSpec_countf3i(spec)
            specCntpi = widget.getSpec_countpi(spec)
            specCntli = widget.getSpec_countli(spec)
            specCntei = widget.getSpec_countei(spec)
            specCnt = (specCntf1i + specCntf2i + specCntf3i + specCntpi + specCntli + specCntei)

            // fill widget only for counted species
            if (specCnt > 0) {
                specArea!!.addView(widget)
                val iName = widget.getSpec_name(spec)
                indivs = individualsDataSource!!.getIndividualsByName(iName!!)
                for (indiv in indivs) {
                    iwidget = ListIndividualWidget(this, null)

                    // load the individuals data
                    iwidget.setIndividual(indiv)
                    specArea!!.addView(iwidget)

                    // show individual notes only when provided
                    val rwidget = ListIndivNoteWidget(this, null)
                    val tRem: String? =
                        if (iwidget.getIndNotes(indiv) == null) ""
                        else iwidget.getIndNotes(indiv)
                    if (tRem!!.isNotEmpty()) {
                        rwidget.setRem(indiv)
                        specArea!!.addView(rwidget)
                    }
                }
            }
        }

        val lwidget = ListLineWidget(this, null)
        specArea!!.addView(lwidget)
    }
    // end of loadData()

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. 
        if (item.itemId == android.R.id.home) {
            super.finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        // close the data sources
        headDataSource!!.close()
        sectionDataSource!!.close()
        countDataSource!!.close()
        individualsDataSource!!.close()

        if (awakePref) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    @Deprecated("Deprecated in Java")
    @SuppressLint("ApplySharedPref", "MissingSuperCall")
    override fun onBackPressed() {
        headDataSource!!.close()
        sectionDataSource!!.close()
        countDataSource!!.close()
        individualsDataSource!!.close()

        NavUtils.navigateUpFromSameTask(this)
        @Suppress("DEPRECATION")
        super.onBackPressed()
    }

    companion object {
        private const val TAG = "ListSpeciesAct"
    }

}